ext/crates/fp/src/vector/generic.rs

```diff
@@ -0,0 +1,434 @@
+use crate::limb::{self, Limb, LimbLength};
+
+use std::marker::PhantomData;
+
+use super::base_generic::{BaseVectorMutP, BaseVectorP};
+use super::internal::InternalBaseVectorP;
+
+/// An `FpVectorP` is a vector over $\mathbb{F}_p$ for a fixed prime, implemented using const
+/// generics. Due to limitations with const generics, we cannot constrain P to actually be a prime,
+/// so we allow it to be any u32. However, most functions will panic if P is not a prime.
+///
+/// Interally, it packs entries of the vectors into limbs. However, this is an abstraction that
+/// must not leave the `fp` library.
+#[derive(Debug, Hash, Eq, PartialEq, Clone)]
+pub struct FpVectorP<const P: u32> {
+    pub(crate) limbs: Vec<Limb>,
+    pub(crate) len: LimbLength<P>,
+}
+
+/// A SliceP is a slice of an FpVectorP. This immutably borrows the vector and implements Copy
+#[derive(Debug, Copy, Clone)]
+pub struct SliceP<'a, const P: u32> {
+    pub(crate) ptr: *const Limb,
```

Comment:
Why does this have to be a raw pointer?

Reply:
So right now `Slice` and `SliceMut` store a pointer to the beginning of the vector in the ambient `FpVector`, so I totally agree that we could store references like the current code does.

However, in the future we'll want to have multiple `SliceMut`s on the same `FpVector` (for tiling, for example). If they all point to the same location, calling `_limbs_mut()` will create multiple mutable slices that all alias each other, and that's instant UB even if we're careful about only accessing the values we care about.

At that point I would like to make it so that the pointer actually points only to the beginning of the region containing the entries we care about. That way, calling `_limbs_mut()` will actually create mutable slices that are next to each other in memory but don't overlap.

Reply:
Wouldn't it be safe to have `&mut [Limb]` in that case? The point is that you check it is non-aliasing when you construct the `SliceMut` itself, instead of every time you use the `SliceMut`.

Reply:
That's a good point. I was ~~stealing~~ using the code for the `rulinalg` crate, and it uses raw pointers. I'll try to replace them by slices and see if it goes wrong. I'm thinking Rust might complain that we can't store a reference to a slice that we just created ourselves. I think storing a slice directly makes the struct a DST, which is awful. Maybe a `Box<[Limb]>`?

Reply:
Fixed by [7fe0f9f](https://github.com/SpectralSequences/sseq/pull/53/commits/7fe0f9ff77b3308a14b088145ade4eda9d367857)